tests/virtualization/livemigration/live_migration_src.pm

```diff
@@ -0,0 +1,279 @@
+    }
+    assert_script_run("virsh undefine $vm_name");
+}
+
+sub run {
+    select_console 'root-ssh';
+    #    cleanup_for_rerun if (check_var('HOST_INSTALL', 0));
+    if (check_var('HOST_INSTALL', 0)) {
+        cleanup_for_rerun;
+    } else {
+        upload_logs('/var/log/zypp/history');
+    }
+    $dst_ip = get_job_autoinst_vars(get_parents->[0])->{SUT_IP};
+    record_info("SRC: $sut_ip");
+    record_info("DST: $dst_ip");
+    #Mount with sync to make sure any write be flushed to the server before the system call returns control
+    assert_script_run("mount -o sync $sut_ip:/tmp/vm_images $images_path") unless get_var("NFS_PATH");
+    assert_script_run("rm -rf $images_path/guestip; mkdir $images_path/guestip");
+    assert_script_run("mkdir -p $images_path/vm-configs");
+    assert_script_run("mkdir -p $images_path/test_state");
+    if (is_sle('<12-sp4')) {
+        zypper_ar($inst_src->{"sles-12-sp4"}, name => "curl_12sp4");
+        zypper_call("up libcurl4");
+    }
+    my $guests = get_var_array('GUESTS');
+    #start all guests
+    for my $guest (@{$guests}) { start_guest($guest, $vms); }
+
+    assert_script_run('if [ -d /root/.ssh ]; then rm -rf /root/.ssh; fi');
+    assert_script_run('ssh-keygen -t rsa -q -f "/root/.ssh/id_rsa" -N ""');
+
+    mutex_wait 'dst_ready';
+
+    assert_script_run("ssh-keyscan $dst_ip > /root/.ssh/known_hosts");
+    assert_script_run("curl -s -k --user root:$testapi::password -T .ssh/id_rsa.pub sftp://$dst_ip/root/.ssh/authorized_keys  --ftp-create-dirs");
+    assert_script_run("ssh root\@$dst_ip mount -o sync $sut_ip:/tmp/vm_images $images_path") unless get_var("NFS_PATH");
+    #clear ring buffer on src and dst hosts
+    assert_script_run("dmesg -C");
+    assert_script_run("ssh root\@$dst_ip dmesg -C");
+
+    my $n = keys %$vms;
+    my $time_cont = 0;
+    #Check each vm if their have got IP address. If they have, their install or stratup are supposed to complete and ready for migration test.
+    until ($n == 0) {
+        while (my ($vm_name, $vm_info) = each %$vms) {
+            next if ($vm_info->{finish} == 1);
+            #check if vm write its IP address in $images_path/guestip/$vm_name on src host.
+            if (script_run("[[ -f $images_path/guestip/$vm_name ]]") == 0) {
+                record_info("$vm_name: ready");
+                migrate_test($vm_name, $vm_info);
+                cleanup($vm_name, $vm_info);
+                $vm_info->{finish} = 1;
+                $n--;
+            } else {
+                die("$vm_name startup time out after " . $time_cont * 2 . " minutes") if ($time_cont == 40);
+                my $vm_state = script_output("virsh domstate --domain $vm_name");
+                if ($vm_state eq "shut off") {
+                    assert_script_run("virsh start $vm_name");
+                    record_info("Reboot: $vm_name install");
+                }
+            }
+        }
+        sleep 120 if ($n != 0);
+        $time_cont++;
```

Comment:
@tonyyuan1  The real vm creation process(from no vm until creation is done and os is up in guest) seems to span in several functions -- start_vm, create_vm and this run function. It will make the debug even harder. I understand it maybe more efficient, but sacrificing the code simplicity and modulerization, it may not worthwhile. And it is not commonly choosen by programmers based on what I have seen for years. I really strongly suggest you abondon this way.
Besides the migration related settings and operations are blended with vm creation too much. I also hope to suggest that we'd better separate vm creation and vm migration.

Reply:
We talked about this offline :).